### PR TITLE
monad: remove S.bind and S.then in favour of R.chain compatibility

### PR DESCRIPTION
**Pull request by davidchambers**
Is this a step in the right direction do you think, @buzzdecafe?

### CODE REVIEW

**Comment by buzzdecafe** [comment]
do you have any interest in fabtasy-land compat? then you would rename `bind` to chain and not worry about native `bind`

File: index.js [hunk 1]
```diff
@@ -55,6 +56,11 @@
   }
   extend(Nothing, Maybe);
 
+  //  Nothing#bind :: (a -> m b) -> m b
+  Nothing.prototype.bind = function(f) {  // jshint ignore:line
+    return this;
+  };
```
Comment (buzzdecafe) [inline index.js]: this is the same as `map`, so you could use a reference to the same function if you were so inclined. not a big deal.

File: index.js [hunk 2]
```diff
@@ -79,6 +85,11 @@
   }
   extend(Just, Maybe);
 
+  //  Just#bind :: (a -> m b) -> m b
+  Just.prototype.bind = function(f) {
+    return f(this.value);
+  };
+
```
Comment (buzzdecafe) [inline index.js]: this is the same as map, so you could use a reference to the same function if you were so inclined. not a big deal.

**Comment by davidchambers** [comment]
> do you have any interest in fantasy-land compat? then you would rename `bind` to chain and not worry about native `bind`

Ah, so *that's* what fantasy-land's `chain` does! Great idea, @buzzdecafe. The only question is whether the exported function should be named `S.bind` or `S.chain`. I prefer the former: Haskell's `>>=` is pronounced "bind", and "chain" only makes sense when used as a method (e.g. `m.chain(f).chain(g)`). It would mean a function named "bind" delegates to a method named "chain", but I *think* I'm okay with that.

File: index.js (hunk 2 above)
Comment (davidchambers) [inline index.js]: I don't mind a bit of repetition if it saves readers from scrolling.

**Comment by benperez** [comment]
:dromedary_camel:

**Comment by buzzdecafe** [comment]
for the record, `ramda.chain` also dispatches to `chain` "methods"

**Comment by davidchambers** [comment]
> for the record, `ramda.chain` also dispatches to `chain` "methods"

In that case `S.bind` is an :alien:! Perhaps the best course of action is to remove `S.bind` and `S.then` and use `R.chain` instead. :)

**Comment by buzzdecafe** [comment]
:+1:

**Pull request merged**